share/wake/lib/core/map.wake

```diff
@@ -0,0 +1,548 @@
+#   ```
+#   mnew scmp | mexists "a" = False
+#   mnew scmp | minsert "a" 1 | mexists "a" = True
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mdelete "b" | mexists "b" = False
+#   ```
+export def mexists (key: k) (map: Map k v): Boolean =
+    isSome (mlookup key map)
+
+
+# Divide the key-value pairs comprising one map into two according to some predicate.
+#
+# *Returns:*
+#   `Pair trues falses` where `trues` contains all the values for which `fn`
+#   returned `True` and `falses` where it returned `False`.  Both resulting maps
+#   use the same key-comparison function as the original `map`.
+export def msplitBy (fn: k => v => Boolean) (map: Map k v): Pair (Map k v) (Map k v) =
+    def cmpKey = map.getMapComparison
+    def pairFn (Pair k v) = fn k v
+    def Pair trues falses = tsplitBy pairFn map.getMapData
+    Pair (Map cmpKey trues) (Map cmpKey falses)
+
+# Discard any key-value pairs in the map for which the predicate fails.
+#
+# *Examples:*
+#   ```
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfilter (\k\_ isVowel k) | mlookup "a" = Some 1
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfilter (\k\_ isVowel k) | mlookup "b" = None
+#   ```
+export def mfilter (fn: k => v => Boolean) (map: Map k v): Map k v =
+    def pairFn (Pair k v) = fn k v
+    editMapData (tfilter pairFn) map
+
+
+# Collect all key-value associations in either of two maps into a single one.
+# If the same key occurs in both, the value from `left` is kept and the one from
+# `right` is discarded.  However, if the key comparison function differs between
+# the two inputs, then the one from the `right` is used.
+#
+# *Examples:*
+#   ```
+#   def left  = listToMap scmp ("a" → 1, "b" → 2, Nil)
+#   def right = listToMap scmp ("b" → 11, "f" → 15, Nil)
+#
+#   munion left right | mlookup "a" = Some 1
+#   munion left right | mlookup "f" = Some 15
+#   munion left right | mlookup "b" = Some 2
+#   ```
+export def munion (left: Map k v) (right: Map k v): Map k v =
+    def Map _ leftTree = left
+    # tunion uses the comparison function of the right tree, so similarly attach
+    # the raw key comparison function of the right map to the output
+    def Map cmpKey rightTree = right
+    tunion leftTree rightTree
+    | Map cmpKey
+
+# Collect all key-value associations in maps with the given conflict resolultion.
+# If the key comparison function differs between the two inputs, then the one
+# from the `right` is used.
+#
+# For a similar function which preserves the original value contained in the
+# map, see `munionWith`.
+#
+# *Examples:*
+#   ```
+#   def left  = listToMap scmp ("a" → 1, "b" → 2, Nil)
+#   def right = listToMap scmp ("b" → 11, "f" → 15, Nil)
+#
+#   munionWith (\_\lv\rv lv + rv) left right | mlookup "a" = Some 1
+#   munionWith (\_\lv\rv lv + rv) left right | mlookup "f" = Some 15
+#   munionWith (\_\lv\rv lv + rv) left right | mlookup "b" = Some 13
+#   ```
+export def munionWith (fn: k => v => v => v) (left: Map k v) (right: Map k v): Map k v =
```

Comment:
The value type should be allowed to change to a v2 here I think?

Reply:
Anything which only occurs in either the left or the right but not both isn't going to trigger the function, so the type needs to remain the same, unless we add a function taking one less argument to cover those cases.